sled-agent/src/storage_manager.rs

```diff
@@ -0,0 +1,451 @@
+    vnic_id_allocator: IdAllocator,
+}
+
+impl StorageWorker {
+    // Idempotently ensure the named dataset exists as a filesystem with a UUID.
+    //
+    // Returns the UUID attached to the ZFS filesystem.
+    fn ensure_dataset_with_id(fs_name: &str) -> Result<Uuid, Error> {
+        Zfs::ensure_filesystem(&fs_name, Mountpoint::Legacy)?;
+        // Ensure the dataset has a usable UUID.
+        if let Ok(id_str) = Zfs::get_oxide_value(&fs_name, "uuid") {
+            if let Ok(id) = id_str.parse::<Uuid>() {
+                return Ok(id);
+            }
+        }
+        let id = Uuid::new_v4();
+        Zfs::set_oxide_value(&fs_name, "uuid", &id.to_string())?;
+        Ok(id)
+    }
+
+    // Formats a partition within a zpool, starting a zone for it.
+    // Returns the UUID attached to the underlying ZFS partition.
+    //
+    // For now, we place all "expected" datasets on each new zpool
+    // we see. The decision of "whether or not to actually use the
+    // dataset" is a decision left to both the bootstrapping protocol
+    // and Nexus.
+    //
+    // If we had a better signal - from the bootstrapping system - about
+    // where Cockroach nodes should exist, we could be more selective
+    // about this placement.
+    async fn initialize_partition(
+        &self,
+        pool: &mut Pool,
+        partition: &PartitionInfo<'static>,
+    ) -> Result<Uuid, Error> {
+        let name = format!("{}/{}", pool.info.name(), partition.name);
+
+        info!(&self.log, "Ensuring dataset {} exists", name);
+        let id = StorageWorker::ensure_dataset_with_id(&name)?;
+
+        info!(&self.log, "Creating zone for {}", name);
+        let zone = ensure_running_zone(
+            &self.log,
+            &self.vnic_id_allocator,
+            partition,
+            &name,
+        )
+        .await?;
+
+        info!(&self.log, "Created zone with address {}", zone.address());
+        pool.add_zone(id, zone);
+        Ok(id)
+    }
+
+    // Small wrapper around `Self::do_work_internal` that ensures we always
+    // emit info to the log when we exit.
+    async fn do_work(&mut self) -> Result<(), Error> {
+        self.do_work_internal()
+            .await
+            .map(|()| {
+                info!(self.log, "StorageWorker exited successfully");
+            })
+            .map_err(|e| {
+                warn!(self.log, "StorageWorker exited unexpectedly: {}", e);
+                e
+            })
+    }
+
+    async fn do_work_internal(&mut self) -> Result<(), Error> {
+        info!(self.log, "StorageWorker creating storage base zone");
+        // Create a base zone, from which all running storage zones are cloned.
+        Zones::create_storage_base(&self.log)
+            .map_err(|e| Error::BaseZoneCreation(e))?;
+        info!(self.log, "StorageWorker creating storage base zone - DONE");
+
+        let mut nexus_notifications = FuturesOrdered::new();
+
+        loop {
+            tokio::select! {
```

Comment:
Would a while on `nexus_notifications.next()` work the same way here instead of select?  

This do_work_internal() should never need to exit, correct?

Reply:
Maybe it's the new_pools_rx.recv here that could be looped, I'm not sure what the nexus_notifications part is doing.